core/include/drivers/i2c.h

```diff
@@ -0,0 +1,346 @@
+ * @i2c: Output reset controller reference upon success
+ *
+ * Return TEE_SUCCESS in case of success
+ * Return TEE_ERROR_DEFER_DRIVER_INIT if reset controller is not initialized
+ * Return TEE_ERROR_ITEM_NOT_FOUND if the resets property does not exist
+ * Return a TEE_Result compliant code in case of error
+ */
+static inline TEE_Result i2c_dt_get_dev(const void *fdt, int nodeoffset,
+					struct i2c_dev **i2c_dev)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	*i2c_dev = dt_driver_device_from_parent(fdt, nodeoffset, DT_DRIVER_I2C,
+						&res);
+	return res;
+}
+#else
+
+static inline TEE_Result i2c_write(struct i2c_dev *i2c_dev __unused,
+				   const uint8_t *buf __unused,
+				   int len __unused)
+{
+	return TEE_ERROR_NOT_SUPPORTED;
+}
+
+static inline TEE_Result i2c_read(struct i2c_dev *i2c_dev __unused,
+				  uint8_t *buf __unused, int len __unused)
+{
+	return TEE_ERROR_NOT_SUPPORTED;
+}
+
+static inline TEE_Result i2c_smbus_raw(struct i2c_dev *i2c_dev __unused,
+				       uint8_t dir __unused,
+				       uint8_t proto __unused,
+				       uint8_t cmd_code __unused,
+				       uint8_t *buf __unused,
+				       unsigned int len __unused)
+{
+	return TEE_ERROR_NOT_SUPPORTED;
+}
+
+static inline TEE_Result i2c_dt_get_dev(const void *fdt __unused,
+					int nodeoffset __unused,
+					struct i2c_dev **i2c_dev)
+{
+	*i2c_dev = NULL;
+	return TEE_ERROR_NOT_SUPPORTED;
+}
+#endif
+
+/**
+ * i2c_dt_get_func - Typedef of function to get reset controller from
+ * devicetree properties
+ *
+ * @a: Pointer to devicetree description of the reset controller to parse
+ * @data: Pointer to data given at i2c_dt_register_provider() call
+ * @res: Output result code of the operation:
+ *	TEE_SUCCESS in case of success
+ *	TEE_ERROR_DEFER_DRIVER_INIT if reset controller is not initialized
+ *	Any TEE_Result compliant code in case of error.
+ *
+ * Returns a struct i2c pointer pointing to a reset controller matching
+ * the devicetree description or NULL if invalid description in which case
+ * @res provides the error code.
+ */
+typedef struct i2c_dev *(*i2c_dt_get_func)(struct dt_driver_phandle_args *a,
+					   void *data, TEE_Result *res);
+
+/**
+ * i2c_dt_register_provider - Register a reset controller provider
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of the reset controller
+ * @get_dt_i2c: Callback to match the reset controller with a struct i2c
+ * @data: Data which will be passed to the get_dt_i2c callback
+ * Returns TEE_Result value
+ */
+static inline
+TEE_Result i2c_register_provider(const void *fdt, int nodeoffset,
+				 i2c_dt_get_func get_dt_i2c, void *data)
+{
+	int subnode = -1;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	res = dt_driver_register_provider(fdt, nodeoffset,
+					  (get_of_device_func)get_dt_i2c,
+					  data, DT_DRIVER_I2C);
+	if (res)
+		return res;
+
+	fdt_for_each_subnode(subnode, fdt, nodeoffset)
+		dt_driver_maybe_add_probe_node(fdt, subnode);
```

Comment:
The inline description above should state that registering an I2C bus device makes its child nodes to be probed.